File: cpp/graph/strongly_connected_component/KosarajuSharir.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class KosarajuSharir {
public:
    vector<vector<int>> search(vector<VertexNode *> &graphs) {
        vector<bool> visited(graphs.size(), false);
        stack<int> stack;
        search(graphs, 0, visited, stack);

        vector<VertexNode *> inverseGraphs(0);
        for (int i = 0; i < (int) graphs.size(); ++i) {
            inverseGraphs.emplace_back(new VertexNode(graphs[i]->value));
        }
        for (int i = graphs.size() - 1; i >= 0; --i) {
            EdgeNode *edge = graphs[i]->next;
            while (edge) {
                int edgeIndex = edge->index;
                EdgeNode *edgeNode = new EdgeNode(i, edge->weight);
                edgeNode->next = inverseGraphs[edgeIndex]->next;
                inverseGraphs[edgeIndex]->next = edgeNode;
                edge = edge->next;
            }
        }

        visited.assign(graphs.size(), false);
        vector<vector<int>> result(0);
        vector<int> aux(0);
        while (!stack.empty()) {
            int index = stack.top();
            stack.pop();
            if (!visited[index]) {
                reverseSearch(inverseGraphs, index, visited, aux);
                result.emplace_back(aux);
                aux.clear();
            }
        }
        return result;
    }

private:
    void search(vector<VertexNode *> &graphs, int start, vector<bool> &visited, stack<int> &stack) {
        visited[start] = true;
        EdgeNode *edge = graphs[start]->next;
        while (edge) {
            int edgeIndex = edge->index;
            if (!visited[edgeIndex]) {
                search(graphs, edgeIndex, visited, stack);
            }
            edge = edge->next;
        }
        stack.push(start);
    }

    void reverseSearch(vector<VertexNode *> &graphs, int start, vector<bool> &visited, vector<int> &aux) {
        visited[start] = true;
        aux.emplace_back(start);

        EdgeNode *edge = graphs[start]->next;
        while (edge) {
            int edgeIndex = edge->index;
            if (!visited[edgeIndex]) {
                reverseSearch(graphs, edgeIndex, visited, aux);
            }
            edge = edge->next;
        }
    }
};

File: py/graph/strongly_connected_component/KosarajuSharir.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class KosarajuSharir:
    def search(self, graphs: List[VertexNode]) -> List[List[int]]:
        visited, stack = [False for _ in range(len(graphs))], []
        self.__search(graphs, 0, visited, stack)

        inverseGraphs = [VertexNode(_.value) for _ in graphs]
        for i in range(len(graphs) - 1, -1, -1):
            edge = graphs[i].next
            while edge:
                edgeIndex, edgeNode = edge.index, EdgeNode(i, edge.weight)
                edgeNode.next, inverseGraphs[edgeIndex].next, edge = inverseGraphs[edgeIndex].next, edgeNode, edge.next

        visited, result, aux = [False for _ in range(len(graphs))], [], []
        while stack:
            index = stack.pop()
            if not visited[index]:
                self.__reverseSearch(inverseGraphs, index, visited, aux)
                result.append(list(aux))
                aux.clear()
        return result

    def __search(self, graphs: List[VertexNode], start: int, visited: List[bool], stack: List[int]) -> None:
        visited[start], edge = True, graphs[start].next
        while edge:
            edgeIndex = edge.index
            if not visited[edgeIndex]:
                self.__search(graphs, edgeIndex, visited, stack)
            edge = edge.next
        stack.append(start)

    def __reverseSearch(self, graphs: List[VertexNode], start: int, visited: List[bool], aux: List[int]) -> None:
        visited[start] = True
        aux.append(start)

        edge = graphs[start].next
        while edge:
            edgeIndex = edge.index
            if not visited[edgeIndex]:
                self.__reverseSearch(graphs, edgeIndex, visited, aux)
            edge = edge.next
